src/main.cpp

```diff
@@ -1,6 +1,224 @@
+#include <cassert>
+#include <cctype>
+#include <cstring>
+#include <ctype.h>
 #include <iostream>
+#include <stdint.h>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#define DEBUG_MODE		   /* empty */
+
+typedef double tl_Number;
+
+enum TokenType {
+	PLUS,	// +
+	MINUS,	// -
+	MULT,	// *
+	DIV,	// /
+	MOD,	// %
+	NUM,	// [0-9]+(.[0-9]+)?
+	STRING, // '[^']' | "[^"]"
+	TRUE,	// true
+	FALSE,	// false
+	LET,	// let
+	NIL,	// nil
+	_EOF,	// '/0'
+	ERROR,	// unknown
+	NUM_TOKENS
+};
+
+// clang-format off
+const std::string TokenNames[TokenType::NUM_TOKENS] = {
+	"PLUS",	
+	"MINUS",	
+	"MULT",	
+	"DIV",	
+	"MOD",	
+	"NUM",	
+	"STRING", 
+	"TRUE",	
+	"FALSE",	
+	"LET",	
+	"NIL",
+	"END_OF_FILE",	
+	"<ERROR>"
+};
+// clang-format on
+
+const static std::unordered_map<std::string, TokenType> keywords = {
+#define KW(str, tokentype) {#str, TokenType::tokentype},
+
+	// clang-format off
+    KW(true , TRUE )
+    KW(false, FALSE)
+    KW(let  , LET  )
+    KW(nil  , NIL  )
+	// clang-format on
+#undef KW
+};
+
+struct Location {
+	std::size_t	  start	 = 0;
+	std::size_t	  length = 0;
+	std::size_t	  line	 = 0;
+	std::uint32_t column = 0;
+
+	std::string_view view(std::string& source) {
+		return std::string_view(source.c_str() + start, length);
+	}
+};
+
+struct Token {
+	TokenType type	   = TokenType::ERROR;
+	Location  location = {};
+
+	std::string_view raw(std::string& source) {
+		return location.view(source);
+	}
+};
+
+std::ostream& operator<<(std::ostream& os, const Token& token) {
+	return os << TokenNames[(int)(token.type)];
+}
+
+class Lexer {
+  private:
+	std::string	  fileName;
```

Comment:
Why is this even needed? Loading text from a file is not the responsibility of this type.

Reply:
To report errors, we store the fileName in the Lexer, parser as well as the VM.  So when it cries, it can do:

```
line 3 in <filename>: Non-terminated string literal.
```

Should I remove it ?